apollo-federation/src/sources/connect/json_selection/apply_to.rs

```diff
@@ -7,12 +7,17 @@ use apollo_compiler::collections::IndexMap;
 use apollo_compiler::collections::IndexSet;
 use itertools::Itertools;
 use serde_json_bytes::json;
-use serde_json_bytes::Map;
+use serde_json_bytes::Map as JSONMap;
 use serde_json_bytes::Value as JSON;
 
 use super::helpers::json_type_name;
+use super::immutable::InputPath;
+use super::lit_expr::LitExpr;
+use super::methods::lookup_arrow_method;
 use super::parser::*;
 
+pub(super) type VarsWithPathsMap<'a> = IndexMap<String, (&'a JSON, InputPath<JSON>)>;
```

Comment:
This is an internal data type that we can easily change. We could go either way, but with `String` I don't have to worry about `&str` lifetimes. I don't think the total number of these `String` values will be large (`"$this"`, `"$args"`, `"$"`, and that's about it).

Reply:
Oh I forgot that this is the container for the variables. We should probably have this be a struct with fields (or an enum, like @dylan-apollo recommended) for each variable that we care about.

Reply:
I looked into using an `enum` for the variable names (rather than `String`), and I think it works/helps! https://github.com/apollographql/router/pull/5762/commits/c2c130c0ad15f8af8bd23fb9837b39446ddeafb3

However, I think we still want an `IndexMap` here, because it's possible some of the `$this`, `$args`, or `$context` variables might not be defined in all circumstances, because they don't all make sense (for example, suppose there are no field arguments to refer to, so no `$args`).

In other words, the `IndexMap` is limited to known variable names now, and parsing fails when you try to use a totally unknown variable name, but it's still a runtime error if you try to use `$args.whatever` when `$args` was never provided to `apply_with_vars`.

Reply:
As a side note, it was my hope that the `&'a JSON` references would copy in constant time, and the `InputPath<JSON>` should also copy fairly quickly because it uses `Rc` internally.

Is that accurate? Is there a better way to make shallow-copying this `IndexMap` cheap?

Reply:
Looks pretty cheap to me! Definitely not worth trying to optimize further unless we see any heat in performance tests. Pretty sure we're at the "things are fast enough, let's focus on features" stage, still.

Luckily, we now have an automated performance test suite in the works for connectors, so we should add some arrow method stuff to that as well to set benchmarks.